### DISCUSSION TITLE
Get Turbine Layout

### DISCUSSION THREAD

**Discussion by AJP-47** [NatLabRockies/floris#887] [Q&A]
Hi,

I'm trying to get the turbine layout using fmodel.get_turbine_layout(). (https://nrel.github.io/floris/_autosummary/floris.floris_model.html#floris.floris_model.FlorisModel.get_turbine_layout).

But I'm getting a type error. I could use this method in Floris3, but it doesn't work in Floris4. Is there any special syntax I should be using?

Thanks

**Comment by misi9170** [comment]
Hi @AJP-47 ,

Could you provide the code you are running? The following code
```python

from floris import FlorisModel

fmodel = FlorisModel("inputs/gch.yaml")

fmodel.set(layout_x=[0, 500.0], layout_y=[0.0, 0.0])

print(fmodel.get_turbine_layout())
print(fmodel.get_turbine_layout(True))
```
produces 
```
(array([  0., 500.]), array([0., 0.]))
(array([  0., 500.]), array([0., 0.]), array([90., 90.]))
```
as expected for me

**Reply by AJP-47** [reply]
Hi @misi9170 ,

Thanks for the quick response. I copy pasted your code and it works fine. I created a minimal working example to recreate the error and that works fine too. So, now I'm stumped as my issue still persists. I guess its the way I'm creating my farm layouts. 

The error I'm getting is 

  File "...\floris4\lib\site-packages\floris\floris_model.py", line 1511, in get_turbine_layout
    xcoords, ycoords, zcoords = self.core.farm.coordinates.T
  File "...\floris4\lib\site-packages\floris\core\farm.py", line 465, in coordinates
    np.array([x, y, z]) for x, y, z in zip(
TypeError: 'numpy.float64' object is not iterable

which suggests that it sees my layout as a single float instead of a list or an array.

Thanks

**Reply by misi9170** [reply]
Hi @AJP-47,

Thanks for the details. I suspect you're right---I think what may be happening here is that, for some reason, `FlorisModel.core.farm.hub_heights` is a float rather than a list or array. I changed this part of the code not long ago in #846 , and it's possible I introduced a new bug in fixing a different one---but looking at what these lines looked like before the change, a float for `Farm.hub_heights` would still have caused an error.

If I'm right about this (that it's `hub_heights` that is a float value rather than a list or array), then there's still the problem of why how `hub_heights` was set as a float in the first place. Is there any indication as to why that might be in your code?

**Reply by rafmudaf** [reply]
Hi @AJP-47 it would be helpful to see your script and your input files so that we can run your exactly configuration on our end. Could you clean up your script (make it as minimal as possible while still hitting the bug) and attach it and your input file to your original post?

**Reply by AJP-47** [reply]
Hi @misi9170,
Thanks for the response. I've passed the hub height as a list, array, int and float and I don't think that is the problem.

@rafmudaf 
I've gone ahead and tried to trace the error and it seems like it depends on the way I get the farm power with get_farm_power().

I've attached some basic code with the input file from the inputs section of the FLORIS examples. In my code there are 2 options in the main.py file. 

- get_farm_power() called before visualize_cut_plane() ==> get_turbine_layout() results in my error (comment out lines 34,35 in main.py)
- get_farm_power() called after visualize_cut_plane()    ==> get_turbine_layout() works fine (comment out lines 18,19 in main.py)

[recreate_err.zip](https://github.com/NREL/floris/files/15096914/recreate_err.zip)

Maybe I'm not supposed to call the functions that way? Either way, I think I have workaround i.e. calculate AEP after I use the visualize function.

**Reply by rafmudaf** [reply]
Thank you, this is very helpful. After reading through your script, I deleted most of it and arrived at a simple example that exposes the bug:

```python

from floris import FlorisModel

fmodel = FlorisModel("inputs/gch.yaml")
fmodel.run()

print(fmodel.get_turbine_layout())
```

**Reply by rafmudaf** [reply]
I think the issue is [here](https://github.com/NREL/floris/blob/main/floris/core/farm.py#L468):

```python
    @property
    def coordinates(self):
        return np.array([
            np.array([x, y, z]) for x, y, z in zip(
                self.layout_x,
                self.layout_y,
                self.hub_heights if len(self.hub_heights.shape) == 1 else self.hub_heights[0,0]
            )
        ])
```

In the if-else statement, `self.hub_heights` is a float when it's shape has length 1, but it should be a list of floats equal to the length of `layout_x`.

**Reply by rafmudaf** [reply]
Scratch that - @misi9170 pointed out that it's in the other part of the if-statement: `self.hub_heights[0,0]` should be `self.hub_heights[0]`.

**Reply by AJP-47** [reply]
Nice! Thanks for digging into it. Anything else I need to do? Like closing the discussion or marking the answer?

**Reply by misi9170** [reply]
Thank you for finding this! I've opened a pull request to fix it, see #890

**Reply by misi9170** [reply]
> Nice! Thanks for digging into it. Anything else I need to do? Like closing the discussion or marking the answer?

@AJP-47 now that we have fixed the bug with #890, we're going to make a minor release to v4.0.1. Once we've done that, we'd appreciate it if you could pull down FLORIS again and confirm that the fix works for you---and then yes, feel free to close the discussion!

**Reply by AJP-47** [reply]
The fix works for me. Thanks!
